Logger instrumentation (#4063)
firecracker-microvm/firecracker: docs/tracing.md

> # Tracing

Author: Added an entry to the change log.

> spent in frequently-called methods, whereas code instrumentation can
> report the exact number of time each method is invoked.

The focus on tracing in Firecracker is to improve debug-ability.

## Implementation

Firecracker implements instrumentation based tracing via
[`log`](https://github.com/rust-lang/log) and
[`log_instrument`](https://github.com/JonathanWoollett-Light/log-instrument),
outputting a `Trace` level log when entering and exiting every function.

It is disabled by default at compile-time, meaning the tracing
functionality has no impact on the release binary.

To enable tracing in Firecracker, compile with `--features tracing` e.g.
`cargo build --features tracing`.

This will result in an increase in the binary size (~100kb) and a
significant regression in performance (>10x). To mitigate the
performance impact you can filter the tracing output as described in the
next section.

## Filtering

You can filter tracing output both at run-time and compile-time.
This can be used to mitigate the performance impact of logging many
traces.

Run-time filtering is implemented with the `/logger` API call, this can

> significantly mitigate the impact on execution time but cannot mitigate

Author: See https://github.com/firecracker-microvm/firecracker/pull/4063#discussion_r1363733136

> the impact on memory usage.

Author: I've expanded it
```text
Run-time filtering is implemented with the `/logger` API call, this can
significantly mitigate the impact on execution time but cannot mitigate
the impact on memory usage. Execution time impact is mitigated by
avoiding constructing and writing the trace log, it still needs to check
the condition in every place it would output a log. Memory usage impact
is not mitigated as the instrumentation remains in the binary unchanged.
```

Compile-time filtering is a manual process assisted by the
[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)

> tool, this can almost entirely mitigate the impact on execution time and

Author: Done.

the impact on memory usage.

### Run-time

You can filter by module path and/or file path at runtime, e.g.:

```bash
curl -X PUT --unix-socket "${API_SOCKET}" \
    --data "{
        \"level\": \"Trace\",
        \"module\": \"api_server::request\",
    }" \
    "http://localhost/logger"
```

Instrumentation logs are `Trace` level logs, at runtime the level must
be set to `Trace` to see them. The module filter applied here ensures
only logs from the `request` modules within the `api_server` crate will
be output.

This will mitigate most of the performance regression.

### Compile-time

> Specific environments can restrict run-time configuration, in these

Author: I'm not sure how, do you have a suggestion?

> Specific environments can restrict run-time configuration, in these

Reviewer: I would just replace commas with full stops.

> Specific environments can restrict run-time configuration, in these

Author: > For this compile-time filtering must be used

Doesn't make much sense alone without the previous context.

> Specific environments can restrict run-time configuration, in these

Reviewer: Can be two sentences then, with the last one being compound. It's certainly a nit though.

> Specific environments can restrict run-time configuration, in these

Author: Okay, made this change.

environments it becomes necessary to support targeted tracing without
run-time re-configuration, for this compile-time filtering must be used.

To reproduce the same filtering as run-time at compile-time, you can use
[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)
at compile-time like:

```bash

> # Install clippy-tracing

Author: Added full stops to all of them.

cargo install clippy-tracing --features log
# Remove all instrumentation.
clippy-tracing --action strip --path ./src
# Adds instrumentation to the specific file/s.
clippy-tracing --action fix --path ./src/api_server/src/request --cfg-attr "feature = \"tracing\""
# Build Firecracker
cargo build --features tracing
```

Then at run-time:

```bash
curl -X PUT --unix-socket "${API_SOCKET}" \
    --data "{
        \"level\": \"Trace\",
    }" \
    "http://localhost/logger"
```

The instrumentation has been stripped from all files other than those at

> `./src/api_server/src/request` so we do not need to apply a run-time

Author: Expended the comment:

```text
The instrumentation has been stripped from all files other than those at
`./src/api_server/src/request` so we do not need to apply a run-time
filter. Runtime filter could be applied but in this case it yields no
additional benefit.
```

filter.

## Example

In this example we start Firecracker with tracing then make a simple API
call.

### API call

```

> jcawl@u08fff4fc3ef156:~/Projects/firecracker$ sudo curl -X GET --unix-socket "/run/firecracker.socket" "http://localhost/"

Reviewer: nit: here and below: can anonymise the prompt.

Author: I've removed `jcawl@u08fff4fc3ef156:`

impact on the release binary.

To enable tracing in Firecracker, add instrumentation with:

```
clippy-tracing \
--action fix \
--path ./src \
--exclude vmm_config/logger.rs,virtio/gen,bindings.rs,net/gen,benches,logger/,signal_handler.rs,time.rs
```

and re-compile with `--features tracing`:

```
cargo build --features tracing
```

This will result in an increase in the binary size (~100kb) and a
significant regression in performance (>10x). To mitigate the
performance impact you can filter the tracing output as described in the
next section.

## Filtering

You can filter tracing output both at run-time and compile-time.
This can be used to mitigate the performance impact of logging many
traces.

Run-time filtering is implemented with the `/logger` API call, this can
significantly mitigate the impact on execution time but cannot mitigate
the impact on memory usage. Execution time impact is mitigated by
avoiding constructing and writing the trace log, it still needs to check
the condition in every place it would output a log. Memory usage impact
is not mitigated as the instrumentation remains in the binary unchanged.

Compile-time filtering is a manual process using the

> [`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)

Author: I've updated the link. Outside of this, yes.

# Tracing

## Introduction

Firecracker implements a framework for instrumentation based tracing.

Instrumentation based tracing as defined by
[Sheng Liang on usenix.org](https://www.usenix.org/legacy/publications/library/proceedings/coots99/full_papers/liang/liang_html/node9.html):

> There are two ways to obtain profiling information: either statistical
> sampling or code instrumentation. Statistical sampling is less
> disruptive to program execution, but cannot provide completely
> accurate information. Code instrumentation, on the other hand, may be
> more disruptive, but allows the profiler to record all the events it
> is interested in. Specifically in CPU time profiling, statistical
> sampling may reveal, for example, the relative percentage of time
> spent in frequently-called methods, whereas code instrumentation can
> report the exact number of time each method is invoked.

The focus on tracing in Firecracker is to improve debug-ability.

> Enabling tracing adds logs output at each function entry and exit.

Reviewer 2: ```suggestion
Enabling tracing adds logs at entry and exit of each function.
``````

# Tracing

> 

Reviewer 3: I'd start this with a brief introduction into what tracing is/what its useful for here. Maybe also settle who the audience is. This seems like a tool that's mainly useful for maintainers, as it requires knowing what files/modules the code sits in/require recompilation.

Author: I've added an introduction
```
Firecracker implements a framework for instrumentation based tracing.

Instrumentation based tracing as defined by [Sheng Liang on usenix.org](https://www.usenix.org/legacy/publications/library/proceedings/coots99/full_papers/liang/liang_html/node9.html):

> There are two ways to obtain profiling information: either statistical sampling or code instrumentation. Statistical sampling is less disruptive to program execution, but cannot provide completely accurate information. Code instrumentation, on the other hand, may be more disruptive, but allows the profiler to record all the events it is interested in. Specifically in CPU time profiling, statistical sampling may reveal, for example, the relative percentage of time spent in frequently-called methods, whereas code instrumentation can report the exact number of time each method is invoked.

The focus on tracing in Firecracker is to improve debugability.
```

Reviewer 3: Mh, this covers the "why is it useful" part, but not really the "what does it do" part, e.g. it assumes the reader already knows what "instrumentation based tracing" is. I think we can start with a description of what happens if we enable tracing. Some example log outputs maybe, and after that go into the details of why its useful (not sure we really need the comparison to `perf`, but I am also not opposed to having it)

Author: I've added an example at the end of the doc.

Reviewer 3: I still think there should be a brief description of it here at the beginning tbh. Something like "Enabling Firecracker's tracing feature will add additional logging statements to each function entry- and exit-point, printing out the time passed between them. This allows the user to debug problems such as deadlocks or high latencies by quickly identifying which function calls are elongated. For an example, see the end of this document."

Author: I've added to the introduction with
```text
Enabling tracing adds logs output at each function entry and exit.
This especially supports debugging problems that relate to deadlocks or
high latencies by quickly identifying elongated function calls.
```

## Introduction

> Instrumentation based tracing is implemented with
> [`log`](https://github.com/rust-lang/log) and
> [`log_instrument`](https://github.com/JonathanWoollett-Light/log-instrument),
> outputting a `Trace` level log when entering and exiting every function.

Reviewer 3: Nit: Use an active voice, e.g. 
```suggestion
Firecracker implements Instrumentation based tracing via
[`log`](https://github.com/rust-lang/log) and
[`log_instrument`](https://github.com/JonathanWoollett-Light/log-instrument),
outputting a `Trace` level log when entering and exiting every function.
```

Author: Updated.

> It is disabled by default at compile-time, as such the
> **tracing functionality has no impact on the release binary.**

Reviewer 3: nit: 
```suggestion
It is disabled by default at compile-time, meaning the
tracing functionality has no runtime or memory impact on the release binary.
```

Reviewer 3: Mh, in this case maybe just say "performance", as that can refer to both execution time and memory?

Author: If we specify `performance` we are saying it may have impacts other than performance. E.g. it might change a log message.

Reviewer 3: That makes sense :)

To enable tracing, remove `"max_level_debug"` from the features
for `log` under [`../src/vmm/Cargo.toml`](../src/vmm/Cargo.toml):

```diff
- log = { version = "0.4.17", features = ["std", "serde", "max_level_debug"] }
+ log = { version = "0.4.17", features = ["std", "serde"] }

> ```

Reviewer 3: Could we do this via a firecracker feature instead? E.g. if "log-tracing" feature of firecracker is disabled, add `max_level_debug`, otherwise omit it

Author: This can be done, it would require a feature on `vmm` like
```
[features]
omit_trace_logs = ["log/max_level_debug"]
```
then a feature on `firecracker` like
```
[features]
default = ["omit_trace_logs"]
omit_trace_logs = ["vmm/omit_trace_logs"]
```
so compiling `firecracker` with `cargo build --no-default-features` would enable trace logs.

This is a bit messy but provides a nicer interaction.

It would be simplified a little if/when we move `vmm` into `firecracker`.

Reviewer 3: Sounds good, I'd be in favor of doing that :)

Author: Okay, I've implemented this.

> This will result in a massive increase in the binary size and a
> massive regression in performance. To mitigate the performance affects
> you can use filtering.

Reviewer 3: nit:
```suggestion
This will result in a significant increase in binary size and
regression in performance. To mitigate the performance impact
you can filter the tracing output as described in the next section.
```

Author: I was switching between `massive` and `significant`, I liked `massive` since it makes people less likely to ask `how significant` and simply informs them its too big to be used in production. I think `significant` can be used with some ballpark percentages.

Reviewer 3: I like the added information on how large the impacts are :)

## Filtering

> 

Reviewer 3: I'd put a short introduction into the two filtering options at the beginning here, something like 

"Firecracker supports filtering tracing output both at run- and compile-time. This can be used to partially mitigate the performance impact of logging thousands of traces. The former allows you to enable tracing on a per-module via API-calls. The latter statically removes tracing from specific module, potentially resulting in better performance at the cost of decreased flexibility."

Author: I've added the text
```
You can filter tracing output both at run-time and compile-time.
This can be used to partially mitigate the performance impact of logging
many traces.

Run-time filtering is implemented with the `/logger` API call, this can
significantly mitigate the performance impact but cannot mitigate the
memory impact.

Compile-time filtering is a manual process assisted by the
[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)
tool, this can almost-entirely mitigate both the performance and memory
impact.
```

Reviewer 3: Mhh, "mitigate the performance impact" is repeated quite a few times here. Can we say "performance impact" in the first paragraph, "execution time" in the second, and "execution time and memory" in the third?

Author: What about
```
You can filter tracing output both at run-time and compile-time.

Run-time filtering is implemented with the `/logger` API call, this can
significantly mitigate the performance impact but cannot mitigate the
memory impact.

Compile-time filtering is a manual process assisted by the
[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)
tool, this can almost-entirely mitigate both the performance and memory
impact.
```

Author: I've reworded it
```
You can filter tracing output both at run-time and compile-time.
This can be used to mitigate the performance impact of logging many
traces.

Run-time filtering is implemented with the `/logger` API call, this can
significantly mitigate the impact on execution time but cannot mitigate
the impact on memory usage.

Compile-time filtering is a manual process assisted by the
[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)
tool, this can almost-entirely mitigate the impact on execution time and
the impact on memory usage.
```

### Run-time

You can filter by module and/or file at runtime, e.g.:

```bash
curl -X PUT --unix-socket "${API_SOCKET}" \
    --data "{
        \"level\": \"Trace\",

>         \"file\": \"src/firecracker/src/api_server_adapter.rs\",

Reviewer 3: Is it possible to do this by rust module path instead of file path instead?

Author: See https://github.com/firecracker-microvm/firecracker/pull/4063#discussion_r1355023059

    }" \
    "http://localhost/logger"

> ```

Author: I've updated the call to show this
```bash
curl -X PUT --unix-socket "${API_SOCKET}" \
    --data "{
        \"level\": \"Trace\",
        \"file\": \"src/api_server/src/request/\",
        \"mdoule\": \"api_server::request\",
    }" \
    "http://localhost/logger"
```

> ```

Reviewer 3: Mhh, is there anything that "file" level filtering can do that module level filtering cannot do?

> ```

Author: Honestly, I cannot think of an example.

> ```

Author: I've removed file filtering.

- Instrumentation logs are `Trace` level, at runtime the level must be
  set to `Trace` to see them.

> - This will only output logs from `src/firecracker/src/api_server_adapter.rs`.

Reviewer 3: nit: I'd turn these pulled points into a paragraph

Author: I've reworded them in a paragraph.

This will mitigate most of the performance regression.

### Compile-time

It can be valuable to produce a binary with targetted tracing that does

> not need to be filtered at run-time.

Reviewer 3: I'd start with a stronger motivation here:

"If the performance impact of runtime filtering is too significant, it is possible to statically exclude specific traces. [...]"

Author: I've re-worded it:
```
Specific environments can restrict run-time configuration, in these
environments it becomes necessary to support targeted tracing without
run-time re-configuration, for this compile-time filtering must be used.
```

To reproduce the same filtering as run-time at compile-time, you can use
[`clippy-tracing`](https://github.com/JonathanWoollett-Light/clippy-tracing)
at compile-time like:

```bash
# Remove all instrumentation.
clippy-tracing --action strip --path ./src
# Adds instrumentation to the specific file/s.

> clippy-tracing --action fix --path ./src/firecracker/src/api_server_adapter.rs

Reviewer 3: I'd add here that you'll need to recompile firecracker after

Author: Added this with
```
# Re-compile Firecracker
./tools/devtool build
```

```

Then at run-time:

```bash
curl -X PUT --unix-socket "${API_SOCKET}" \
    --data "{
        \"level\": \"Trace\",
    }" \
    "http://localhost/logger"
```

- Instrumentation logs are `Trace` level, at runtime the level must be
  set to `Trace` to see them.

> - There is now no need to add a run-time filter.

Reviewer 3: Again I'd write this as a paragraph instead of bullet points

Author: I've reworded them in a paragraph.
